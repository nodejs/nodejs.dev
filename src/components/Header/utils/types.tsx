export enum HeaderType { Link, Anchor }

export interface LinkHeaderProps {
  type: HeaderType,
  to: string,
  classes: string,
  title: string,
  activeClassName: string,
  partiallyActive: boolean
}

export interface AnchorHeaderProps {
  type: HeaderType,
  href: string,
  classes: string,
  title?: string,
  mobileTitle?: string,
  checkMobile?: boolean
}
