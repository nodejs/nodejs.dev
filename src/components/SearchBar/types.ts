import type { SerialisedIndexData } from 'elasticlunr';
import type { SearchResult } from '../../types';

export interface SearchProps {
  localSearchLearnPages: {
    index: SerialisedIndexData<SearchResult>;
  };
}
