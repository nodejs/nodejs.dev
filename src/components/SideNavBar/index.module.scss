@import '../../containers/Navigation/index.module.scss';
@import '../NavigationSection/index.module.scss';
@import '../NavigationItem/index.module.scss';

.side-nav {
  @extend %navigation;

  &__open {
    @extend %navigation__open;
  }

  &--fixed {
    @extend %navigation--fixed;

    height: 100vh;
    top: calc(var(--nav-height) * 2);
    touch-action: none;
    width: 100%;
  }

  &--fixed &__open {
    top: 1.8rem;
  }

  .side-nav-section {
    @extend %navigation-section;

    padding: 0 var(--space-20);

    &__title {
      @extend %navigation-section__pure-title;

      margin-left: 0;

      b {
        color: var(--color-brand-primary);
        font-weight: var(--font-weight-semibold);
      }
    }

    .side-nav-item {
      @extend %navigation-item;

      padding: calc(var(--space-20) / 2) 0 calc(var(--space-20) / 2)
        var(--space-12);

      &::before {
        box-shadow: inset 0 0 0 2px var(--color-border-accent);
        line-height: 2.6rem;
        top: 1rem;
      }

      &::after {
        content: none;
      }

      &:hover::before {
        box-shadow: inset 0 0 0 2.4rem var(--color-border-accent) !important;
      }

      &--active {
        color: var(--color-text-accent);
        font-weight: 600;

        &::before {
          box-shadow: inset 0 0 0 2.4rem var(--color-border-accent);
        }
      }
    }
  }
}
