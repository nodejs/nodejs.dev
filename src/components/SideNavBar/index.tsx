import React, { useRef, useState } from 'react';
import { SideNavBarItem } from '../../types';
import NavigationItem from '../NavigationItem';
import '../../styles/about.scss';

// eslint-disable-next-line no-shadow
export enum SideNavBarKeys {
  about = 'about',
  governance = 'governance',
  community = 'community',
  workingGroups = 'working-groups',
  releases = 'releases',
  resources = 'resources',
  trademark = 'trademark',
  privacy = 'privacy',
  security = 'security',
  packageManager = 'download/package-manager',
}

const sideNavBarItems: SideNavBarItem[] = [
  {
    title: 'About',
    slug: SideNavBarKeys.about,
  },
  {
    title: 'Project Governance',
    slug: SideNavBarKeys.governance,
  },
  {
    title: 'Community',
    slug: SideNavBarKeys.community,
  },
  {
    title: 'Working Groups',
    slug: SideNavBarKeys.workingGroups,
  },
  {
    title: 'Releases',
    slug: SideNavBarKeys.releases,
  },
  {
    title: 'Resources',
    slug: SideNavBarKeys.resources,
  },
  {
    title: 'Trademark Policy',
    slug: SideNavBarKeys.trademark,
  },
  {
    title: 'Privacy Policy',
    slug: SideNavBarKeys.privacy,
  },
  {
    title: 'Security Reporting',
    slug: SideNavBarKeys.security,
  },
  {
    title: 'Package Manager',
    slug: SideNavBarKeys.packageManager,
  },
];

// eslint-disable-next-line no-shadow
export enum OverflowTypes {
  unset = 'unset',
  hidden = 'hidden',
}

interface NavBarProps {
  pageKey: string;
}

export default function SideNavBar({ pageKey }: NavBarProps): JSX.Element {
  const [navOpen, setNavOpen] = useState<boolean>(false);
  const toggle = (): void => setNavOpen(!navOpen);

  const className = navOpen
    ? 'side-nav-about side-nav--open'
    : 'side-nav-about';

  if (typeof document !== 'undefined') {
    document.body.style.overflow = navOpen
      ? OverflowTypes.hidden
      : OverflowTypes.unset;
  }

  const navElement = useRef<HTMLElement | null>(null);

  return (
    <nav className={className} ref={navElement}>
      <button type="button" className="side-nav__open" onClick={toggle}>
        Menu
      </button>
      <ul className="community-nav__list">
        {sideNavBarItems.map(({ title: commTitle, slug }) => {
          return (
            <NavigationItem
              key={slug}
              title={commTitle}
              isLearn={false}
              isRead={false}
              isActive={slug === pageKey}
              slug={slug}
              baseUrl="/"
            />
          );
        })}
      </ul>
    </nav>
  );
}
