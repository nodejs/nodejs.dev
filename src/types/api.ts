export interface ApiChange {
  version: string | string[];
  'pr-url': string;
  description: string;
}

export interface ApiUpdate {
  type: 'added' | 'removed' | 'deprecated' | 'introduced_in' | 'napiVersion';
  version: string[];
}

export interface ApiComponentData {
  type?: string;
  name?: string;
  source_link?: string;
  update?: ApiUpdate;
  changes?: ApiChange[];
}

export interface ApiType {
  name: string;
  slug: string;
}
