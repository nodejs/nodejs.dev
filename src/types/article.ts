import { PageTableOfContents } from './markdown';

export interface ArticleData {
  body: string;
  tableOfContents: PageTableOfContents;
  frontmatter: {
    title: string;
    description: string;
    displayTitle?: string;
  };
  fields: { authors: string[]; slug?: string };
}
