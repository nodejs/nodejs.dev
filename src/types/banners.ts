export interface BannersIndex {
  endDate: string;
  link: string;
  text?: string;
  html?: string;
  startDate: string;
}

export interface BannersBLM {
  link: string;
  text: string;
  visible: boolean;
}

export interface Banners {
  banners: {
    index: BannersIndex;
    blacklivesmatter: BannersBLM;
  };
}
