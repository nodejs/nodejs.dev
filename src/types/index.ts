export interface HomepageData {
  page: {
    id: string;
    html: string;
    frontmatter: {
      title: string;
      displayTitle: string;
      subTitle: string;
      description: string;
      learnLinkText: string;
      beginnerGuideHeaderText: string;
      beginnerGuideBodyText: string;
      doMoreWithNodeHeaderText: string;
      doMoreWithNodeBodyText: string;
      nodeFeatureHeader1: string;
      nodeFeatureHeader2: string;
      nodeFeatureHeader3: string;
      nodeFeature1: string;
      nodeFeature2: string;
      nodeFeature3: string;
      nodeFeatureAltText: string;
    };
  };
}

export interface LearnPageContext {
  slug: string;
  relativePath: string;
  next: PaginationInfo;
  previous: PaginationInfo;
  navigationData: NavigationSectionData;
}

export interface LearnPageData {
  doc: {
    id: string;
    html: string;
    tableOfContents: string;
    frontmatter: { title: string; description: string };
    fields: { authors: string[] };
  };
}

export interface PaginationInfo {
  slug: string;
  title: string;
}

export interface NavigationSectionItem {
  slug: string;
  title: string;
  section: string;
}

export interface NavigationSectionData {
  [index: string]: NavigationSectionItem[];
}

/**
 * Payload on the `stickychange` custom event
 */
export interface StickyChange {
  stuck: boolean;
  target: HTMLElement;
}

export interface SentinelObserverSetupOptions {
  container: HTMLElement;
  stickyElementsClassName: string;
  root?: HTMLElement | null;
  headerRootMargin?: string;
  footerRootMargin?: string;
}

declare global {
  interface Window {
    previousPath: string;
  }
}

export interface Page {
  editPath?: string;
  data: {
    page: {
      html: string;
      tableOfContents: string;
      frontmatter: {
        title: string;
        description: string;
      };
      fields: { authors: string[] };
    };
  };
  location?: Location;
}
export interface CommunityNavigationSection {
  title: string;
  sections: string[];
}

export interface BlogPostAuthor {
  id?: string;
  name: string;
  url: string;
}

export interface BlogMetaData {
  node: {
    frontmatter: { title: string; author: BlogPostAuthor[] };
    fields: { date: string; slug: string };
  };
}
export interface BlogPostsList {
  blogs: {
    edges: BlogMetaData[];
  };
}

export interface BlogPageData {
  blog: {
    html: string;
    excerpt: string;
    frontmatter: { title: string; author: BlogPostAuthor[] };
    fields: { slug: string; date: string };
  };
  recent: {
    edges: BlogMetaData[];
  };
}

export interface BlogPageContext {
  slug: string;
  relativePath: string;
  next: PaginationInfo;
  previous: PaginationInfo;
  navigationData: NavigationSectionData;
}
