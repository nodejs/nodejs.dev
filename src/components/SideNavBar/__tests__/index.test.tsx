import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import SideNavBar, { SideNavBarKeys } from '..';
import { SideNavBarItem } from '../../../types';

const navbarButtons = [
  {
    pageKey: SideNavBarKeys.about,
    expectedText: 'About Node.js',
    expectedHref: '/about/',
  },
  {
    pageKey: SideNavBarKeys.governance,
    expectedText: 'Project Governance',
    expectedHref: '/about/governance/',
  },
  {
    pageKey: SideNavBarKeys.releases,
    expectedText: 'Releases',
    expectedHref: '/about/releases/',
  },
  {
    pageKey: SideNavBarKeys.resources,
    expectedText: 'Resources',
    expectedHref: '/about/resources/',
  },
  {
    pageKey: SideNavBarKeys.privacy,
    expectedText: 'Privacy Policy',
    expectedHref: 'https://privacy-policy.openjsf.org/',
  },
  {
    pageKey: SideNavBarKeys.security,
    expectedText: 'Security Reporting',
    expectedHref: '/about/security/',
  },
  {
    pageKey: SideNavBarKeys.getInvolved,
    expectedText: 'Get Involved',
    expectedHref: '/get-involved/',
  },
  {
    pageKey: SideNavBarKeys.collabSummit,
    expectedText: 'Collab Summit',
    expectedHref: '/get-involved/collab-summit',
  },
  {
    pageKey: SideNavBarKeys.contribute,
    expectedText: 'Contribute',
    expectedHref: '/get-involved/contribute',
  },
  {
    pageKey: SideNavBarKeys.codeOfConduct,
    expectedText: 'Code of Conduct',
    expectedHref:
      'https://github.com/nodejs/node/blob/main/doc/contributing/code-of-conduct.md',
  },
  {
    pageKey: SideNavBarKeys.download,
    expectedText: 'Download',
    expectedHref: '/download/',
  },
  {
    pageKey: SideNavBarKeys.packageManager,
    expectedText: 'Package Manager',
    expectedHref: '/download/package-manager/',
  },
  {
    pageKey: SideNavBarKeys.previousReleases,
    expectedText: 'Previous Releases',
    expectedHref: '/download/releases/',
  },
];

describe('SideNavBar', () => {
  it('renders correctly', () => {
    const { container } = render(<SideNavBar pageKey={SideNavBarKeys.about} />);
    expect(container).toMatchSnapshot();
  });

  navbarButtons.forEach(button => {
    it(`should contain a href to '~${button.expectedHref}'`, () => {
      render(<SideNavBar pageKey={button.pageKey} />);
      const aboutNavBarElement = screen.getByText(button.expectedText);
      expect(aboutNavBarElement.getAttribute('href')).toBe(button.expectedHref);
    });
  });

  it('should set the a single page as active', () => {
    const { container } = render(
      <SideNavBar pageKey={SideNavBarKeys.releases} />
    );
    const innerHtml = container.innerHTML;
    const activeLinks = innerHtml.match('navigationItemActive');
    expect(activeLinks && activeLinks.length).toBe(1);
  });

  it('should contain a href to https link', () => {
    const title = 'title';
    const slug = 'https://test.domain';
    const items: SideNavBarItem[] = [{ title, slug, isTitle: true }];
    render(<SideNavBar pageKey={SideNavBarKeys.releases} items={items} />);
    const navBarElement = screen.getByText(title);
    expect(navBarElement.getAttribute('href')).toBe(slug);
  });

  it('should close the Navigation when a link is clicked', async () => {
    const activeClassName = 'navigationItemActive';
    render(<SideNavBar pageKey={SideNavBarKeys.releases} />);
    const releasesNavBarElement = screen.getByText('Releases');
    expect(releasesNavBarElement.classList).toContain(activeClassName);
    await userEvent.click(releasesNavBarElement);
    expect(releasesNavBarElement.classList).toEqual(
      expect.not.arrayContaining([activeClassName])
    );
  });
});
