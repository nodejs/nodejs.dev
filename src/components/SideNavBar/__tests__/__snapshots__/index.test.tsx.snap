// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavBar renders correctly 1`] = `
<div>
  <nav
    class="sideNav"
  >
    <button
      class="sideNavOpen"
      type="button"
    >
      Menu
    </button>
    <div
      class="sideNavSection"
    >
      <a
        aria-current="page"
        class="t-body2 navigationItem navigationItemActive"
        href="/about/"
        id="link-about"
      >
        About
      </a>
      <a
        class="t-body2 navigationItem"
        href="/community/"
        id="link-community"
      >
        Community
      </a>
      <a
        class="t-body2 navigationItem"
        href="/download/package-manager/"
        id="link-download/package-manager"
      >
        Package Manager
      </a>
      <a
        class="t-body2 navigationItem"
        href="/about/privacy/"
        id="link-about/privacy"
      >
        Privacy Policy
      </a>
      <a
        class="t-body2 navigationItem"
        href="/about/governance/"
        id="link-about/governance"
      >
        Project Governance
      </a>
      <a
        class="t-body2 navigationItem"
        href="/about/releases/"
        id="link-about/releases"
      >
        Releases
      </a>
      <a
        class="t-body2 navigationItem"
        href="/resources/"
        id="link-resources"
      >
        Resources
      </a>
      <a
        class="t-body2 navigationItem"
        href="/about/security/"
        id="link-about/security"
      >
        Security Reporting
      </a>
      <a
        class="t-body2 navigationItem"
        href="/about/working-groups/"
        id="link-about/working-groups"
      >
        Working Groups
      </a>
    </div>
  </nav>
</div>
`;
